### PR TITLE
Implemented more error-handling and reconnection.

### PR DESCRIPTION
**Pull request by marpie**
This change is [Reviewable](https://reviewable.io/reviews/jtpereyda/boofuzz/90)

### CODE REVIEW

**Comment by jtpereyda** [comment]
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90#-KT2YxQ0FH7A8_WuyUsj:-KT2YxQ15uEpWu_nts8K:b-j40943) (no related file):*
@marpie How did your use case work out? You said that in your situation a connection was being dropped due to a malformed packet. Is that still the case, or did you encounter something else?

I'm trying to think of all the possible meanings of a dropped connection. What I have so far:

1. Connection deliberately closed. UUT is behaving reasonably. Boofuzz can more or less ignore.
2. Connection dropped due to crash. UUT is crashed but the error should be reported through another means. Boofuzz can more or less ignore.
3. Connection issues. This makes the test invalid but will hopefully be detected when the next test case fails. Boofuzz can more or less ignore.

---

*[boofuzz/sessions.py, line 741 at r1](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90#-KT2YeisOiUmeV53FLh_:-KT2YeisOiUmeV53FLha:bflazz7) ([raw file](https://github.com/jtpereyda/boofuzz/blob/599d7619e5d0444f035a2d0472e46a560f994311/boofuzz/sessions.py#L741)):*
> ```Python
>                 # Try to send payload down-range
>                 self.targets[0].send(data)
>             except socket.error as e:
> ```


@marpie It seems like if we catch these errors in `socket_connection.py` then we don't need it here, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90)*

**Comment by marpie** [comment]
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90#-KT2YxQ0FH7A8_WuyUsj:-KT4BF9kNCXSwHay0oJB:b-xmja6c) (no related file):*
<details><summary><i>Previously, jtpereyda (Joshua Pereyda) wrote…</i></summary>
> @marpie How did your use case work out? You said that in your situation a connection was being dropped due to a malformed packet. Is that still the case, or did you encounter something else?
> 
> I'm trying to think of all the possible meanings of a dropped connection. What I have so far:
> 
> 1. Connection deliberately closed. UUT is behaving reasonably. Boofuzz can more or less ignore.
> 2. Connection dropped due to crash. UUT is crashed but the error should be reported through another means. Boofuzz can more or less ignore.
> 3. Connection issues. This makes the test invalid but will hopefully be detected when the next test case fails. Boofuzz can more or less ignore.
</details>
With the PR both targets I tried to code on worked and produced good crashes that were caught correctly.

Since I haven't had the time to extensively test Boofuzz on more targets I have a limited view on the functionality and the inner workings, I tried to fix it in the right locations and it seems to work now on my targets.

---

*[boofuzz/sessions.py, line 741 at r1](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90#-KT2YeisOiUmeV53FLh_:-KT49qKJ7rKGPZjWfHYp:b-mcx929) ([raw file](https://github.com/jtpereyda/boofuzz/blob/599d7619e5d0444f035a2d0472e46a560f994311/boofuzz/sessions.py#L741)):*
<details><summary><i>Previously, jtpereyda (Joshua Pereyda) wrote…</i></summary>
> @marpie It seems like if we catch these errors in `socket_connection.py` then we don't need it here, right?
</details>
No, the problem is that the connection is not established again after being dropped. I couldn't locate the exact code that should reconnect, therefore I added this code that retries to send the data again and that seems to work in both my targets.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90)*

**Comment by jtpereyda** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[boofuzz/sessions.py, line 741 at r1](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90#-KT2YeisOiUmeV53FLh_:-KTA6Ac7PoHO-mAyqPox:b-o3tn2h) ([raw file](https://github.com/jtpereyda/boofuzz/blob/599d7619e5d0444f035a2d0472e46a560f994311/boofuzz/sessions.py#L741)):*
<details><summary><i>Previously, marpie (Markus Piéton) wrote…</i></summary>
> No, the problem is that the connection is not established again after being dropped. I couldn't locate the exact code that should reconnect, therefore I added this code that retries to send the data again and that seems to work in both my targets.
</details>
I see, my mistake; the socket code happens on recv and this code is for send.

When does this code run for you? Does it re-establish a connection in the middle of a test? If the UUT drops a connection due to bad input, it seems like you would want to restart the whole test instead of connecting and continuing, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90)*

**Comment by jtpereyda** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[boofuzz/sessions.py, line 741 at r1](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90#-KT2YeisOiUmeV53FLh_:-KTaGaiYdkNON587hTmP:b-zedaj4) ([raw file](https://github.com/jtpereyda/boofuzz/blob/599d7619e5d0444f035a2d0472e46a560f994311/boofuzz/sessions.py#L741)):*
<details><summary><i>Previously, jtpereyda (Joshua Pereyda) wrote…</i></summary>
> I see, my mistake; the socket code happens on recv and this code is for send.
> 
> When does this code run for you? Does it re-establish a connection in the middle of a test? If the UUT drops a connection due to bad input, it seems like you would want to restart the whole test instead of connecting and continuing, right?
> 
</details>
@marpie Thanks again for the PR.

The new comment says "... if it was closed in the previous fuzz-case". After further thought, this should not be needed. The connection is `open`ed and `close`d every single fuzz case. See method `_fuzz_current_case`, lines 873 to 891.

However, it might be that you put this here to deal with a connection that was closed _earlier in the same fuzz case_. Is that right?

Thanks! I hope it doesn't seem like nitpicking; I'm just trying to square out the use cases in my mind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90)*

**Comment by jtpereyda** [comment]
@marpie By the way when you make changes, you're welcome to just push them to your fork's feature branch, which will automatically update the PR, or you can keep making new ones; either way.

**Comment by jtpereyda** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[boofuzz/sessions.py, line 741 at r1](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90#-KT2YeisOiUmeV53FLh_:-KTaO6Ku5VDCPgQRAV2t:brliylh) ([raw file](https://github.com/jtpereyda/boofuzz/blob/599d7619e5d0444f035a2d0472e46a560f994311/boofuzz/sessions.py#L741)):*
<details><summary><i>Previously, jtpereyda (Joshua Pereyda) wrote…</i></summary>
> @marpie Thanks again for the PR.
> 
> The new comment says "... if it was closed in the previous fuzz-case". After further thought, this should not be needed. The connection is `open`ed and `close`d every single fuzz case. See method `_fuzz_current_case`, lines 873 to 891.
> 
> However, it might be that you put this here to deal with a connection that was closed _earlier in the same fuzz case_. Is that right?
> 
> Thanks! I hope it doesn't seem like nitpicking; I'm just trying to square out the use cases in my mind.
</details>
@marpie I dove into the code again and realized that `Session._fuzz_current_case` stops sending and receiving once it hits the fuzz node (lines 873 to 891). Therefore, if the UUT drops the connection after fuzzing, Boofuzz should not perform any more sends before closing and re-opening the connection.

Given that understanding, I think we can remove this code. Any objects?

The joys of making generic frameworks...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90)*

**Comment by marpie** [comment]
@jtpereyda I retried that and I think my objective was to re-connect to the UUT after the connection was dropped and while inside a "current" fuzz-case.

How should we proceed with that given that boofuzz doesn't seem to re-connect correctly if the connection is dropped.

A simple test case that lets you reproduce the issue is http://www.exploit-db.com/wp-content/themes/exploit/applications/687ef6f72dcbbf5b2506e80a375377fa-freefloatftpserver.zip with the simple test case: https://gist.github.com/marpie/0f7cdf1e70521b6f43e7db33c00a6783

**Comment by jtpereyda** [comment]
## TL;DR
@marpie I downloaded and ran your test. I described my results below (let me know if your results differ). I also proposed a modified approach in PR #93. If you have time, try it out and let me know how it goes. At least on my machine, the test runs smoother and UUT failures are reported nicely. See [How can I checkout a GitHub pull request?](http://stackoverflow.com/questions/27567846/how-can-i-checkout-a-github-pull-request) if you decide to try it out.

## Test Results
Here's the partial output and error after running your code (with boofuzz on current `master`, commit `3ecff0ffe86859b87aace554477db64d1fff7b70`):

````
[2016-10-11 17:49:58,694] Test Case: 6
[2016-10-11 17:49:58,694]     Info: primitive name: None, type: Delim, default value:  
[2016-10-11 17:49:58,694]     Info: Test case 6 of 47 for this node. 6 of 9999 overall.
[2016-10-11 17:49:58,696]     Received: 32 32 30 20 46 72 65 65 46 6c 6f 61 74 20 46 74 70 20 53 65 72 76 65 72 20 28 56 65 72 73 69 6f 6e 20 31 2e 30 30 29 2e 0d 0a b'220 FreeFloat Ftp Server (Version 1.00).\r\n'
[2016-10-11 17:49:58,696]   Test Step: Fuzzing Node 'user'
[2016-10-11 17:49:58,697]     Transmitting 509 bytes: 55 53 45 52 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 66 74 70 0d 0a b'USER                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    ftp\r\n'
Traceback (most recent call last):
  File "boofuzz-freefloatftp-simple.py", line 119, in &lt;module&gt;
    sess.fuzz()
  File "c:\users\josh\code\boofuzz\boofuzz\sessions.py", line 414, in fuzz
    self._fuzz_current_case(*fuzz_args)
  File "c:\users\josh\code\boofuzz\boofuzz\sessions.py", line 868, in _fuzz_current_case
    self.transmit(target, self.fuzz_node, edge)
  File "c:\users\josh\code\boofuzz\boofuzz\sessions.py", line 734, in transmit
    self.targets[0].send(data)
  File "c:\users\josh\code\boofuzz\boofuzz\sessions.py", line 128, in send
    num_sent = self._target_connection.send(data=data)
  File "c:\users\josh\code\boofuzz\boofuzz\socket_connection.py", line 178, in send
    num_sent = self._sock.send(data)
socket.error: [Errno 10053] An established connection was aborted by the software in your host machine
````

This output suggests that the crash is happening during `send`, not `recv`.

The Errno, 10053, corresponds to `errno.ECONNABORTED`. In my case this was due to a host-based antivirus program shutting down the connection.

After working around that, the application crashes on Test Case 6 and one of two things happens:

 1. If I acknowledge the crash popup that Windows gives me quickly enough, the transmission succeeds, and the application reports an error on the next test case when it can't re-connect.
 2. If I don't click the Windows popup quickly enough, I get an Errno 10054, corresponding to `errno.ECONNRESET`. This suggests an error that should be logged by boofuzz.

All of these situations seem to me like something that boofuzz should report as an error in the process under test.

Note: When the application crashes, the process monitor shoudl catch it, but that isn't happening for me right now.

## Proposal
My proposal for handling these errors is in PR #93 , built off of this PR.

---

Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/jtpereyda/boofuzz/90#-:-KTqMrdnPl_E087Ur0ln:bysap5v)*
<!-- Sent from Reviewable.io -->
